Move the psid-map to softwire in snabb-softwire-v2

tsyesika opened the pull request:
This is the implementation #792.

This PR moves the `psid-map` list in the yang schema over to the `softwire` list. There were many options or routes to implementation however I tried to pick the ones which I thought would keep the code the simplest, fastest with the smallest size footprint with regard to the binding table.

The new softwire schema looks like this:
```yang
list softwire {
   key "ipv4 psid padding";

   leaf ipv4 {
      type inet:ipv4-address;
      mandatory true;
   }

   leaf padding {
      type uint16 { range 0..0; }
      default 0;
   }

   leaf br-address {
      type inet:ipv6-address;
      mandatory true;
   }

   leaf b4-ipv6 {
      type inet:ipv6-address;
      mandatory true;
   }

   leaf psid {
      type uint16;
      mandatory true;
   }

   container port-set {
      leaf psid-length {
        type uint8 { range 0..16; }
        mandatory true;
      }

      leaf shift {
        type uint8 { range 0..16; }
      }

      leaf reserved-ports-bit-count {
      type uint8 { range 0..16; }
      default 0;
      }
   }
}
```
_NB: I've removed the descriptions above but not in the schema itself to make it more readable._

I did want to put the `psid` field in the `port-set` container as I think it makes the most sense however, YANG requires all keys be children of the `list` (section: [7.8.2](https://tools.ietf.org/html/rfc6020#section-7.8.2)).

I decided I'd keep `binding_table.lua` using the separated `psid-map` and `softwire` lists in order to save space and retain the ability to lookup with a ctable via the streaming interface. Changing the schema to the above means: the `softwire` list changes from a ctable to a cltable with the value being a simple lua table. Maybe shift should be `default 0;` and I believe that should prevent that however it wasn't in the original schema (v1). The ctable is desired and I also **don't** want the psid information on the softwire as it's already going to be in the `psid-map`.

The change also means that the `end-addr` option goes away! This also brings up the option of switching away from `rangemap.lua` to a plain-old ctable. I decided to take this route in order to allow me to add a new `psid-map` entry without restarting. That however changes some of the interfaces for example the `lookup` interface.

So what actually happens with these changes?

### (Re-)Starting the lwaftr

In [load](src/apps/lwaftr/binding_table.lua) I iterate over the softwires to build up the `psid-map` ctable and put the softwires in a ctable without the now superfluous information.

### Adding a softwire via the reconfigurable interface

1. We see it's a softwire and hand off to [add_softwire_entry_actions](src/apps/config/support/snabb-softwire-v2.lua) as per usual
2. We create a blob for the ctable
3. We check if there is a psid-map entry for that ctable
4. If not we also make an rpc to our new shiny `add_psid_map_entry` function.
5. We commit those changes as per usual and the binding table adds it to its ctable as per usual too (adding `psid-map` entries is done in a similar fashion to adding softwires).

**Question:** What do we do if they add a softwire with an existing `psid-map` entry however the `psid-map` entry is different from others with the same IP? Do we update the `psid-map` entry changing it for the other softwires or error?

What's left to do is:
- [x] As the `load` function has changed since I wrote and tested the migration, I need to fix that.
- [x] Update the translation layer
- [x] Update the documentation.

tsyesika committed:
Move psidmap onto softwire in snabb-softwire-v2

This moves the psidmap onto the softwire instead of being its own block.
The "pisd" leaf still exists on the softwire directly rather than the
new "port-set" container as I believe all keys must leaves of the list
(section 7.8.2 of RFC6020).

tsyesika committed:
Add migration for psidmap on softwire

This adds a minor version for 3.2.0 (3.2.0.1) which moves the psidmap
onto the softwire. This should migrate all configurations away from
using the psidmap block and remove the "end-addr" parameter for good.

The migration can be run on a file using:

    snabb lwaftr migrate-configuration -f 3.2.0.1 <config file path>

This presumes the configuration file is already at 3.2.0 (with the
br-address leaf-list removed).

tsyesika committed:
Convert lwaftr's data files to 3.2.0.1

This converts the existing conf files for the lwaftr's tests to version
3.2.0.1. This removes the psidmap block and adds it onto the new
"port-set" container on the softwire itself.

tsyesika committed:
Convert v2 migration to handle 3.2.0 and 3.2.0.1

This converts the migration added for 3.2.0.1 to be part of the 3.2.0
snabb-softwire-v2 migration. This handles both adding "br-address" on
softwire (away from leaf-list) and psid-map to softwire.

This will not convert something converted "part way" with a previous
version of the migration. If you've already converted a config away from
the br-address list but not yet to the integrated psid-map on softwire
then you'll need to checkout a prior commit to this one and use the
3.2.0.1 converter.

This also removes the bump version which was only used in development.

tsyesika committed:
Update revision statement to include new port-set

This updates the revision statement to bump the date to when the latest
change occured. It also includes the information about the switch from
using the "psid-map" list over to using the "port-set" container on the
softwire itself. This has been done in the same revision as the
br-address as there has been no release with these changes inplace.

tsyesika committed:
Support new softwire in support and binding table

This adds support in the snabb-softwire-v2 support for the new format
where it formats new softwires as a blob the binding table can
quickly add. It also adds a transformer for initial loads of a config
to the binding table so it's able to consume configs in the new
format.

NB: This does not yet handle adding a softwire with a PSID which is
not yet known to data plane (i.e. does not appear on another
softwire).

tsyesika committed:
Change binding table's psid map to ctable

The new snabb-softwire-v2 config no longer allows for ipv4 address
ranges. This means there is no real reason to continue using the
rangemap system and it can be switched to a plain ctable.

This will give us the ability to add port-set entries on the fly
without restarting the lwaftr (how it was previously done) or
rebuilding the entire rangemap (impractical).

tsyesika committed:
Add ability to add psid entries and update tests

This adds the ability to add and remove psidmaps without restarting the
lwaftr. This allows for the addition of softwires with a new psidmapping
without requiring any additional restart. It also updates the unit tests
for the binding table to reflect the new configuration and some of the
internal changes to it.

There still remains quite a bit to do, this does NOT address:

- The leader's version of the binding table isn't updated
- No longer used versions of the PSID map are not removed
- Validation doesn't occur to check the port-set entries line up with
other port-set entries (what should even happen if they don't?)
- Translation layer isn't updated
- Some tests are failing which needs addressing.

tsyesika committed:
Remove validate_config and validate_update

These might need to come back however right now they only validated that
the softwire that was being added had a corrisponding psid-map. That no
longer applies.

tsyesika committed:
Fix migration to 3.2.0

The migration to convert files pre-snabb-softwire-v2 (3.2.0) has broken
when i changed the binding_table.lua:load function. This now takes an
older version of it and uses that in order to build the binding table so
it can migrate the configurations.

tsyesika committed:
Update snabbvmx tests to the psid-map on softwires

tsyesika committed:
Fix a test to use snabb-softwire-v2 schema

This fixes an `snabb config add` test to use the updated schema. This
means adding the psid-map on the softwire using the new `port-set`
container. Previously it was erroring as it was using the old config.

This also removes the no longer valid test to check adding a softwire
without a `psid-map` entry was invalid as entires are now on the
softwire. Whilst this reduces the number of tests I think there will be
additional tests in the (near) future when the behavour around adding
softwires with different port-set info to another softwire with the same
address is decided.

tsyesika committed:
Fix a few more tests and remove redundent tests

This removes some of the tests to check if softwires without psidmaps
errored as it doesn't make as much sense with the change. It also fixes
a few other little issues with some config tests.

Outstanding: ietf-softwire translation layer is yet to do so those tests
rightly fail.

tsyesika committed:
Fix config translation layer for ietf-softwire

This updates the ietf-softwire translation layer for the
snabb-softwire-v2 configuration now using the `port-set` container
on `softwire` in place of the `psid-map` list.

Commit:
Fix small nits

Commit:
Update documentation for new schema

Commit:
Fix typo nit in docs

Commit:
Remove shift and padding parameters

Commit:
Update test data for removal of shift and padding

Commit:
Update migration for removal of shift and padding

Commit:
Update snabbvmx configs to remove shift too

wingo commented on src/program/lwaftr/tests/data/tunnel_icmp.conf:
```diff
@@ -1,101 +1,124 @@
+      br-address 1e:2:2:2:2:2:2:af;
+      port-set {
+        psid-length 0;
+        reserved-ports-bit-count 16;
```
AFAIU this initial config always has psid-length+ shift = 16 so reserved-ports-bit-count should be 0; no?  This comment applies here and elsewhere.

wingo reviewed (commented):
On the low level: excellent, excellent work.  Nice stuff.  Just a nit regarding what is probably an error in the migrate-configuration thing.

On the high level, have you done any performance testing yet?  I am a little skeptical that there will be an advantage to using a ctable internally for the psid map.  To me it seems like a range map is the right thing; dunno.  Wdyt?  Is there a reason to switch?

xray7224 commented:
I have not yet done performance testing, it's my next job, hopefully shouldn't take long.

The reason for switching psidmap to a ctable over a rangemap is simply so we can add new psid entries without having to rebuild it. It becomes quite difficult to add them otherwise. It seems the rangemap was mainly useful because it allowed for range lookup was it not? We can switch back but I personally think being able to add or remove a psid-map entry is quite valuable however maybe it's not worth it.

wingo commented:
right, you do have to rebuild the table when you add new ips to it.  that happens infrequently though.  relatedly it's hard to know when to rebuild:
 - when you remove a softwire, do you leave its entry in the psid map?
   - if you remove the psid map entry, you could corrupt other softwires on that ip
   - if you keep it you could mistakenly identify that ip as yours for the purpose of hairpinning
   - when you add a softwire and the psid map has a conflicting entry, do you know for sure that that entry is used?

We could add refcounts or something in the psid map but for now I think it's OK to assume that modifying the PSID map is a bit of a PITA and so the data structure change is the least of our worries.

On the perf side I am wary of introducing the ctable overhead here.  Pretty sure the binary search in the rangemap is a win.  I guess you can go ahead with perf testing of this branch as-is and then we can see what the real-world win (if any) of the rangemap is -- but that would definitely be my first suggestion as a perf improvement.

Commit:
Fix migration to handle larger files and add padding leaf

This fixes the migration so it uses less memory, allowing for larger
files to be migrated. It's currently been tested with a 100MB file.

This also adds back the padding leaf as it causes issues in the ctable
streaming lookup of the softwires. The entry must confirm to certain
sizes or multi_copy.dasl used by ctable streaming will assert an error.

Commit:
Convert the binding table to use rangemap

This converts the binding table to use rangemap and reintroduces the
padding parameter to snabb-softwire-v2. The rangemap was reintroduced
over the ctable due to the performance impact, as rangemap uses binary
search to find the port-sets, it's notably faster.

The padding parameter was re-introduced as multi_copy.dasl complains
if the size of a ctable (in this case softwire) is not divisable by 4
when multiplied by the entries_per_lookup (max_displacement + 1).

xray7224 commented:
@wingo PTAL.

I've made some changes as you suggested, switching back to the rangemap helped a lot with performance issues I discovered last week. It does mean we restart if you do happen to add a softwire which defines a new port-set however hopefully that won't be too often. The `padding` parameter still exists too, this is largely related to https://github.com/Igalia/snabb/blob/lwaftr/src/lib/multi_copy.dasl#L57.

I have performance tested this and it can reach 2 MPPS with only 0.000050% packet loss, the interesting parts of the test are as follows:
```
Applying 9.000000 Gbps of load.
  NIC 0:
    TX 1959917 packets (1.959917 MPPS), 1077954350 bytes (8.999939 Gbps)
    RX 1959917 packets (1.959917 MPPS), 1156351030 bytes (9.627112 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  NIC 1:
    TX 1959917 packets (1.959917 MPPS), 1077954350 bytes (8.999939 Gbps)
    RX 1959917 packets (1.959917 MPPS), 999557670 bytes (8.372765 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.200000 Gbps of load.
  NIC 0:
    TX 2003477 packets (2.003477 MPPS), 1101912350 bytes (9.199966 Gbps)
    RX 2003476 packets (2.003476 MPPS), 1182050840 bytes (9.841074 Gbps)
    Loss: 0 ingress drop + 1 packets lost (0.000050%)
  NIC 1:
    TX 2003477 packets (2.003477 MPPS), 1101912350 bytes (9.199966 Gbps)
    RX 2003477 packets (2.003477 MPPS), 1021773270 bytes (8.558854 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.400000 Gbps of load.
  NIC 0:
    TX 2047038 packets (2.047038 MPPS), 1125870900 bytes (9.399998 Gbps)
    RX 2014834 packets (2.014834 MPPS), 1188752060 bytes (9.896865 Gbps)
    Loss: 0 ingress drop + 32204 packets lost (1.573200%)
  NIC 1:
    TX 2047038 packets (2.047038 MPPS), 1125870900 bytes (9.399998 Gbps)
    RX 2024271 packets (2.024271 MPPS), 1032378210 bytes (8.647686 Gbps)
    Loss: 0 ingress drop + 22767 packets lost (1.112192%)
Applying 9.600000 Gbps of load.
  NIC 0:
    TX 2090550 packets (2.090550 MPPS), 1149802500 bytes (9.599806 Gbps)
    RX 2037774 packets (2.037774 MPPS), 1202286660 bytes (10.009546 Gbps)
    Loss: 0 ingress drop + 52776 packets lost (2.524503%)
  NIC 1:
    TX 2090550 packets (2.090550 MPPS), 1149802500 bytes (9.599806 Gbps)
    RX 2090550 packets (2.090550 MPPS), 1066180500 bytes (8.930830 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.800000 Gbps of load.
  NIC 0:
    TX 2134151 packets (2.134151 MPPS), 1173783050 bytes (9.800021 Gbps)
    RX 2037836 packets (2.037836 MPPS), 1202323240 bytes (10.009850 Gbps)
    Loss: 0 ingress drop + 96315 packets lost (4.513036%)
  NIC 1:
    TX 2134151 packets (2.134151 MPPS), 1173783050 bytes (9.800021 Gbps)
    RX 2134151 packets (2.134151 MPPS), 1088417010 bytes (9.117093 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 10.000000 Gbps of load.
  NIC 0:
    TX 2177656 packets (2.177656 MPPS), 1197710800 bytes (9.999796 Gbps)
    RX 2038025 packets (2.038025 MPPS), 1202434750 bytes (10.010779 Gbps)
    Loss: 0 ingress drop + 139631 packets lost (6.411986%)
  NIC 1:
    TX 2177656 packets (2.177656 MPPS), 1197710800 bytes (9.999796 Gbps)
    RX 2177656 packets (2.177656 MPPS), 1110604560 bytes (9.302946 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 10.000000 Gbps of load.
  NIC 0:
    TX 2177658 packets (2.177658 MPPS), 1197711900 bytes (9.999806 Gbps)
    RX 2037978 packets (2.037978 MPPS), 1202407020 bytes (10.010548 Gbps)
    Loss: 0 ingress drop + 139680 packets lost (6.414230%)
  NIC 1:
    TX 2177658 packets (2.177658 MPPS), 1197711900 bytes (9.999806 Gbps)
    RX 2177658 packets (2.177658 MPPS), 1110605580 bytes (9.302955 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.800000 Gbps of load.
  NIC 0:
    TX 2134114 packets (2.134114 MPPS), 1173762700 bytes (9.799851 Gbps)
    RX 2037801 packets (2.037801 MPPS), 1202302590 bytes (10.009679 Gbps)
    Loss: 0 ingress drop + 96313 packets lost (4.513020%)
  NIC 1:
    TX 2134114 packets (2.134114 MPPS), 1173762700 bytes (9.799851 Gbps)
    RX 2134114 packets (2.134114 MPPS), 1088398140 bytes (9.116935 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.600000 Gbps of load.
  NIC 0:
    TX 2090593 packets (2.090593 MPPS), 1149826150 bytes (9.600003 Gbps)
    RX 2037801 packets (2.037801 MPPS), 1202302590 bytes (10.009679 Gbps)
    Loss: 0 ingress drop + 52792 packets lost (2.525217%)
  NIC 1:
    TX 2090593 packets (2.090593 MPPS), 1149826150 bytes (9.600003 Gbps)
    RX 2090593 packets (2.090593 MPPS), 1066202430 bytes (8.931013 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.400000 Gbps of load.
  NIC 0:
    TX 2047035 packets (2.047035 MPPS), 1125869250 bytes (9.399985 Gbps)
    RX 2037809 packets (2.037809 MPPS), 1202307310 bytes (10.009718 Gbps)
    Loss: 0 ingress drop + 9226 packets lost (0.450701%)
  NIC 1:
    TX 2047035 packets (2.047035 MPPS), 1125869250 bytes (9.399985 Gbps)
    RX 2047035 packets (2.047035 MPPS), 1043987850 bytes (8.744934 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
Applying 9.200000 Gbps of load.
  NIC 0:
    TX 2003480 packets (2.003480 MPPS), 1101914000 bytes (9.199980 Gbps)
    RX 2003480 packets (2.003480 MPPS), 1182053200 bytes (9.841094 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
  NIC 1:
    TX 2003480 packets (2.003480 MPPS), 1101914000 bytes (9.199980 Gbps)
    RX 2003480 packets (2.003480 MPPS), 1021774800 bytes (8.558867 Gbps)
    Loss: 0 ingress drop + 0 packets lost (0.000000%)
```

wingo commented:
Perf numbers look good.  I have a few questions:
 * What workload are you running -- 550-byte packets?
 * How many softwires in your test binding table?  Probably we need to test with 2M.
 * How many active softwires in the test workload?  20K is good enough for final perf numbers.
 * What is the difference in the compiled configuration size before and after with a 2M-entry binding table?

It also looks to me like a few side utilities need to be updated, like the binding table generator. (`snabb lwaftr generate-binding-table`)

wingo commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -63,21 +63,18 @@ module(..., package.seeall)
 
 local band, lshift, rshift = bit.band, bit.lshift, bit.rshift
 
 -- FIXME: Pull these types from the yang model, not out of thin air.
```
We can remove this comment, given that the psid map is no longer a YANG thing.

wingo commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -195,11 +192,11 @@ end
--- psid_map_value_t pointer, which has psid_length and shift members.
+-- The IPv4 values are host-endianness uint32 values. The key is a
+-- softwire_key_t which has a "addr" and psid_info is a psid_map_value_t
+-- pointer, which has psid_length and shift members.
```
I think the old comment still applies -- the psid map is a rangemap that will coalesce contiguous ranges of keys that have the same value.  However I think perhaps this function is no longer needed, as I think right now the only user of this function was to write out the psid-map yang list, and that is no more.  Consider removing the function and comment.

wingo commented on src/apps/lwaftr/lwaftr.lua:
```diff
@@ -276,6 +276,12 @@ end
 function LwAftr:remove_softwire_entry(entry_key_blob)
    self.binding_table:remove_softwire_entry(entry_key_blob)
 end
+function LwAftr:add_psid_map_entry(psid_blob)
```
What's this for?  Appears to not be called.

wingo commented on src/apps/lwaftr/rangemap.lua:
```diff
@@ -83,7 +83,7 @@ function RangeMapBuilder:add(key, value)
 end
 
-function RangeMapBuilder:build(default_value)
+function RangeMapBuilder:build(default_value, ignore_duplicates)
```
I am pretty sure that we don't want to ignore duplicates -- rather we want to require that all entries with the same key have the same value.  Maybe that means the builder should use a ctable as its intermediate data store and verify that when a key is added but already present in the builder, that the value is the same.

wingo commented on src/lib/yang/data.lua:
```diff
@@ -349,7 +349,11 @@ local function table_parser(keyword, keys, values, string_key, key_ctype,
       for k,_ in pairs(values) do
          local id = normalize_id(k)
-         value[id] = struct[id]
+         if type(struct[id]) == "cdata" and type(value[id]) == "cdata" then
```
This seems like the wrong thing but I'm not sure.  Is this change needed?

wingo commented on src/program/lwaftr/doc/configuration.md:
```diff
@@ -134,47 +142,42 @@ encapsulated in IPv6 and sent to the AFTR; the AFTR does the reverse.
+    port-set {
+      psid-length 6;
+      reserved-ports-bit-count 0;
+      shift 10;
```
I think this example is out of date; no "shift" parameter any more.

wingo commented on src/program/lwaftr/doc/configuration.md:
```diff
@@ -134,47 +142,42 @@ encapsulated in IPv6 and sent to the AFTR; the AFTR does the reverse.
+`reserved-ports-bit-count` and `shift` in the `port-set` container are
+both optional.
 
 An entry's `psid-length`, `shift` and `reserved-ports-bit-count`
```
Relatedly, this paragraph needs updating.

wingo reviewed (commented):
Some nits.  Looking very good, thank you for your work here!

Commit:
Fix small documentation and other nits

This fixes some of the documentation surrounding the configuration
change. This includes both user facing documentation as well as
commends throughout the code.

It also removes the newly introduced "ignore_duplicates" flag in the
RangeMapBuilder in favour of building a ctable of the added entries in
order to verify that the softwires with the same port-set definition
don't get added multiple times to the psid-map and that they don't
define different values for the port-set.

It removes two now unused functions which were to add or remove psid
map entries. These were removed as the switch back to rangemap makes
it unfeasible to modify the psid-map in the binding table once it's
been built.

xray7224 commented:
So, in answer to your questions:

> 1. What workload are you running -- 550-byte packets?

Yep, I am.

> 2. How many softwires in your test binding table? Probably we need to test with 2M.

I used https://people.igalia.com/dpino/lwaftr/lwaftr-migrated.conf which by my count has 1,000,062 softwires. Is there a configuration with 2M that I can use?

> 3. How many active softwires in the test workload? 20K is good enough for final perf numbers.

Unsure, Is there an easy way of getting the count?

> 4. What is the difference in the compiled configuration size before and after with a 2M-entry binding table?

This is using the above binding table I mentioned in your 2nd question of ~1M softwires.

Before: 93M
After: 166M

So it grew by 73M.

wingo committed:
Fix structural typing of YANG data

YANG data should be structurally typed.  If I have a struct with just a
"uint16_t a" member, then any other struct that has that same shape
should have the same FFI type.  To implement this we have to memoize the
string -> type function so that all types with the same string
representation use the same type object.  However this is a bit tricky,
givent that types can contain members with compound types -- before,
these nested members with compound types were being treated as nominal
fresh types instead of structural types.  This commit fixes that bug.

wingo commented:
Thanks for the perf answers!  In order:

 1. Great.
 2. 1M is fine.  Just making sure it wasn't a test with just like 10 softwires.
 3. I don't know.  Open the pcap in wireshark and see how many packets are in it?  Or tcpdump?  I think these files are 20K active softwires though.
 4. Hmm.  I guess it's OK though given that our perf numbers are stable on this workload.

I think I have a fix for the YANG ctype error in #807; if it works for you, hit merge and it will join this PR.

Commit:
Add test for lwaftr's generate-binding-table subcommand

This adds a test which runs the generate-binding-table command and
verifies that it exits with a 0 status code and produces the expected
output for the command line parameters.

It doesn't check the contents of the block but it does check certain
expected things which should verify some problems if they occur.

wingo committed:
Merge pull request #807 from Igalia/cltype

Fix structural typing of YANG data

wingo commented on src/apps/lwaftr/binding_table.lua:
```diff
@@ -158,9 +155,9 @@ function BindingTable:remove_softwire_entry(entry_key_blob)
-local lookup_key = softwire_key_t()
 function BindingTable:lookup(ipv4, port)
-   local psid = self:lookup_psid(ipv4, port)
+   local lookup_key = ffi.new(self.softwires.entry_type).key
```
This is going to be bad for lookup perf (though we usually use the other interface).  Can we add a FIXME here?

tsyesika replied:
Hm, how about we just create the lookup key when we create the binding table, seems that's what we were doing in the old code. I'm thinking something like this:

```lua
local lookup_key
function BindingTable.new(psid_map, softwires)
   local ret = {
      psid_map = assert(psid_map),
      softwires = assert(softwires),
   }
   lookup_key = ret.softwires.entry_type().key
   return setmetatable(ret, {__index=BindingTable})
end
```
then this can just be reverted to the way it was and I think the code will be effectively the same, right? If I'm missing something more complicated then we can put a FIXME but I'd perfer to fix it now if it's easy.

wingo replied:
Good idea, this looks good to me!

tsyesika replied:
Hokay, i'll fix this and push it in a second :)

Commit:
Fix small nit in generate_binding_table_test.py

There were a few small nits found such as not including the \n when
checking if something ends in a curly brace and an undefined variable
instead of checking the length of the list.

Commit:
Update generate-binding-table for new schema changes

This updates lwaftr's generate-binding-table subcommand to include the
psid-map in the softwire to reflect the changes made in the
snabb-softwire-v2 schema changes.

Commit:
Fix small performance nit

This fixes a small performance nit in binding_table.lua which would
have been caused by allocating the key of the psid-map on every
lookup. This now reverts it back to how it was, creating the key only
when the binding table is initiated.

This should improve the performance of the BindingTable:lookup
interface a little.

Commit:
Update migrations fixing the port-set definition.

Commit:
Fix lwaftr's migrate-configuration.

This fixes lwaftr's mgirate-configuration commands going from older
versions to more recent. There have been some changes throughout the
code which means it hasn't worked but it should now.

It's been tested with:
https://people.igalia.com/dpino/lwaftr/lwaftr.conf

Commit:
Fix lwaftr's test vlan configs

wingo commented:
LGTM.  Let's merge this then then see about merging lwaftr-kristian to lwaftr.
